Create 'useCopyToClipboard' hook

axbarsan opened the pull request:
Fixes giantswarm/giantswarm#7802

axbarsan commented:
Still a WIP, but would love to get some feedback for what I've done so far, before digging into the 3 big components that are left.

ograu commented on src/components/getting-started/fileblock.js:
```diff
@@ -23,110 +23,87 @@ import ReactCSSTransitionGroup from 'react-addons-css-transition-group';
+    e.preventDefault();
 
-    this.setState({ clicked: false });
+    const contentToCopy = Helpers.dedent(children);
```
It would be nice and a good practice to import just the function we are using from `Helpers`. 
I know you found it like this, but since we are refactoring this now, we can quickly improve change this as well.

ograu commented on src/components/getting-started/fileblock.js:
```diff
@@ -23,110 +23,87 @@ import ReactCSSTransitionGroup from 'react-addons-css-transition-group';
+        <div className='content' ref={preElement}>
+          <div className='codeblock--filename'>{fileName}</div>
+          <div className='codeblock--filecontents'>
+            {hideText ? undefined : <Line text={Helpers.dedent(children)} />}
```
We could do this kind of things this way instead, unless we need this `undefined`:

```javascript
{hideText && <Line text={Helpers.dedent(children)} />}
```

oponder commented on src/components/getting-started/codeblock.js:
```diff
@@ -24,115 +24,91 @@ import ReactCSSTransitionGroup from 'react-addons-css-transition-group';
-      .map(function(x) {
+      .map(x => {
         return x.props.children;
       })
```
I'm not much of a code golfer, but I'm happy with these kinds of improvements and nitpicks, and am comfortable reading them.

ograu replied:
I am, kind of obsessed about this, so let me know if I go beyond what's reasonable!

ograu commented:
> 
> Does anyone have ideas on how we can test this in the test suite?

Probably mocking `copy`

axbarsan commented:
> Nice work, great use case for custom hook, this was in my list as well.
> 
> I would suggest that you include in this PR pertinent changes in `V5ClusterDetailsTable` and `V4ClusterDetailsTable`. You will find todos there I left some time ago:
> 
> ```js
> // TODO We are repeating this in several places, refactor this to a reusable HOC / hooks.
> ```

Yes, this was in my first comment. Wanted to get feedback before digging into those, since those are really big.

ograu commented:
I really though that I wrote these comments...

axbarsan commented:
> I think it's a really nice refactor.
> 
> I have a few comments. Will also do some exploratory testing.
> 
> Does anyone have ideas on how we can test this in the test suite?

Could mock the `useCopyToClipboard` file and make a mock hook that returns a `[boolean, jest.fn()]` type of data structure, and then you'd control the behaviour of the component through that, and then test it for different values.

As for testing the actual hook, I'll add that today.

axbarsan commented:
> I really though that I wrote these comments...

I mean I said that in my first pull request comment haha 😬

axbarsan commented on src/components/UI/cluster_id_label.js:
```diff
@@ -37,68 +40,50 @@ const Label = styled.span`
-          {this.props.clusterID}
-        </Label>
+  return (
+    <Wrapper onMouseLeave={() => setCopyToClipboard(clusterID)}>
```
Good catch :D

oponder commented on src/components/getting-started/codeblock.js:
```diff
@@ -24,115 +24,87 @@ import ReactCSSTransitionGroup from 'react-addons-css-transition-group';
+        <div className='codeblock--buttons'>
+          <a
+            href='#'
+            onClick={() => setClipboardContent(null)}
```
We're missing now the `e.preventDefault()`, so clicking this causes a bump up to the top of the page.

You can test by creating a cluster and going down to one of the copy buttons on the getting started guide.

Maybe the whole setup here too weird, probably something i arrived at because of the animation flourishes I wanted to add to this action.

axbarsan replied:
Oh yes, added that now. Thank you 👍

axbarsan commented:
Thank you guys! I'll work on refactoring the other components to use this.

oponder commented:
Closes: https://github.com/giantswarm/giantswarm/issues/7802

axbarsan commented:
Ended up not refactoring the `Users`, `V4ClusterDetailTable` and `V5ClusterDetailTable` components completely, to move quicker.

`Users` is 80% refactored, it was split up in multiple components and the logic was cleaned, it would still need to be converted to functional component, and use react reducers instead of regular state setters to make the code lighter.

As for `V4ClusterDetailTable` and `V5ClusterDetailTable`, the URI block has been taken out and refactored, to use the new `useCopyToClipboard` hook. Nothing else was touched.

Also added the `emotion` babel plugin, to be able to use component selectors ([benefits](https://emotion.sh/docs/babel-plugin-emotion)).

Everything looks and behaves the same 👍

ograu commented:
Adrian I really appreciate your work here, but it is too big a PR in my opinion.
Is there any way to extract what doesn't belong to this PR? And submit it in other PRs?

I am thinking basically in the refactors you are providing.
Thanks

axbarsan commented on src/lib/effects/useCopyToClipboard.js:
```diff
@@ -0,0 +1,22 @@
+    }
+  }, [copiedText]);
+
+  return [copiedText !== null, copyToClipboard];
```
It's not a boolean, it's the the actual copied text, which can be a `string`, or a `null`, the first value of the returned tuple is a boolean (`string !== null` is `true`, and `null !== null` is `false`)

axbarsan commented:
Yes, I did stop refactoring because this PR was getting too large and it was taking up too much time. As it is right now, it's pretty readable imo

oponder commented:
I'll make some time to review this today.  Thx @axbarsan

oponder commented on src/components/GettingStarted/CodeBlock.js:
```diff
@@ -1,133 +1,115 @@
+          </a>
+        </div>
+        <BaseTransition
+          in={this.state.clicked}
```
```suggestion
          in={hasContentInClipboard}
```

axbarsan replied:
I think these appeared while merging master, will fix today

axbarsan replied:
Good catch, thank you!

oponder commented on src/components/GettingStarted/FileBlock.js:
```diff
@@ -1,127 +1,112 @@
+          </a>
+        </div>
+        <BaseTransition
+          in={this.state.clicked}
```
```suggestion
          in={hasContentInClipboard}
```

oponder commented on src/components/Users/Users.js:
```diff
@@ -1,40 +1,36 @@
-import BootstrapTable from 'react-bootstrap-table-next';
 import Button from 'UI/Button';
-import copy from 'copy-to-clipboard';
+import DeleteUserModal from './DeleteUserModal';
```
I like these refactorings, they make the code better! Really appreciate it.

Would've been best in a separate PR though to keep this one focused on the topic of creating that hook.

oponder commented on yarn.lock:
```diff
@@ -701,7 +701,18 @@
   dependencies:
     regenerator-runtime "^0.13.2"
 
+<<<<<<< HEAD
```
Merge conflict here.

If it helps feel free to delete the file and `yarn install` from scratch
